### PR TITLE
Response error handling

### PR DESCRIPTION
**Pull request by davideluque**
Decoding the Base64 response sent by NemID forks into two cases:

- The decoded Base64 is an XML Digital Signature document, which is identified by checking if the decoded string starts with `'<?xml'`
- The decoded Base64 is an error code, for example, `APP001` which encoded is `QVBQMDAx`.

The idea of this implementation is to raise an error whenever the decoded base64 is a NemID error code - or when the response is something else than XMLDSig or NemID Error code.

This implementation creates a class for each NemID error, including a message for developers, a message for users in Danish, and in English. Some classes don't include any message because NemID does not provide it, i.e., all messages have been taken from NemID Error Codes document. 

This implementation will raise the class corresponding to the received error.

For example, if the error is SRV001, then when doing `NemID::Authentication::Response.new(base64_str)`, the code will:

```ruby
raise NemID::Errors::SRV001Error
```
Now, since `SRV001Error` is a subclass of `SRVError`, which at the same time is a subclass of `ResponseError`, developers can recover from any of the three, and receive the same error information (msg, da, en).

Example:

**Recovering from the specific error** `APP002Error` 
```ruby
begin
 response = NemID::Authentication::Response.new(base64_string)
rescue NemID::Errors::APP002Error => e
  puts e    # Developer-friendly error:  "Correct the sign text. [JS Client: See tool at /developers/signtextviewer.jsp]"
  puts e.da # User-friendly error in Danish: "Der er opstået en teknisk fejl. Forsøg igen. Kontakt support, hvis problemet fortsætter."
  puts e.en # User-friendly error in English: "A technical error has occurred. Please try again. Contact support if the problem persists"
end
```

**Recovering from the error category** `APP`
```ruby
begin
 response = NemID::Authentication::Response.new(base64_string)
rescue NemID::Errors::APPError => e
 # the same as in the previous code can be done here..
end
```

**Recovering from** `ResponseError`
```ruby
begin
 response = NemID::Authentication::Response.new(base64_string)
rescue NemID::Errors::ResponseError => e
 # the same as in the previous code can be done here..
end
```

This gives developers the freedom to handle the error that comes from NemID in any way they want, and at the same time giving useful information for them or the users.

### CODE REVIEW

**Comment by davideluque** [comment]
Guide to NemID Error Codes:

[NemID Error Codes.pdf](https://github.com/davideluque/nemid/files/5094711/NemID.Error.Codes.pdf)

**Comment by davideluque** [comment]
AUTH011 msg has a sentence:

> "Thereafter, please log on NemID at [url to Service Provider site containing client(s)]"

This has been changed to

> "Thereafter, please try again."

Since it is very inconvenient to provide an URL or Mobile App name.

**Danish version: Prøv derefter igen.** ??

image

**Comment by davideluque** [comment]
I think this is ready for a first glance @ludvigandersen

File: lib/nemid/authentication/response.rb [hunk 1]
```diff
@@ -39,6 +44,22 @@ def has_rid?
...
+      rescue NameError
+        return false
+      end
+
+      def error(str)
+        klass = "NemID::Errors::#{str}Error"
+        if class_exists?(klass)
+          return eval(klass) 
+        else
+          return NemID::Errors::ResponseError
```
[the remark this answers is not in the document]
Comment (ludvigandersen) [reply lib/nemid/authentication/response.rb:59]: The questions was primarily to get clarification the error handling, just like we discussed over Slack last week.
I think it is fine returning what we are returning right now.

File: lib/nemid/errors/app.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    end      
+
+    class APP008Error < APPError
+      def initialize(msg="Fix the integration issue. [JS Client: See tool at " \
+        "/developers/validateparameters.jsp]")
+        super
+      end
+    end
+
+    class APP009Error < APPError; end
```
Comment (davideluque) [inline lib/nemid/errors/app.rb:45]: Msg: "Invalid HSession"

**Comment by davideluque** [comment]
To-do:

Add information about this to the README.

**Pull request merged**